### ISSUE TITLE
Cannot update translation ota in Android 

### ISSUE THREAD

**Issue by AndroidDeveloperEob** [phrase/phrase-android#72]
last working version: 3.1.6
Implementation in Application (3.1.6):
```
  Phrase.setup(
            this,
            BuildConfig.PHRASE_DISTRIBUTION_ID,
            BuildConfig.PHRASE_SECRET,
            storeLocale
        )
      getCoreSemanticAppVersion()?.let {
            Phrase.setAppVersion(it)
        } ?: ExceptionReporter.logException(
            IllegalStateException("Invalid app version. Cannot get core semantic app version")
        )
        Phrase.updateTranslations()
```
tested versions from 3.5.0 to 3.10.0-rc2
Previous implementation not working:
```
  Phrase.setup(
            this,
            BuildConfig.PHRASE_DISTRIBUTION_ID,
            BuildConfig.PHRASE_SECRET,
            storeLocale
        )
        getCoreSemanticAppVersion()?.let {
            Phrase.setAppVersion(it)
        } ?: ExceptionReporter.logException(
            IllegalStateException("Invalid app version. Cannot get core semantic app version")
        )
        Phrase.updateTranslations()
```

New implementation from 3.7.0:
```
        Phrase.setup(
            this,
            BuildConfig.PHRASE_DISTRIBUTION_ID,
            BuildConfig.PHRASE_SECRET,
            storeLocale
        )
        getCoreSemanticAppVersion()?.let {
            Phrase.setAppVersion(it)
        } ?: ExceptionReporter.logException(
            IllegalStateException("Invalid app version. Cannot get core semantic app version")
        )
        Phrase.updateTranslations(object : TranslationsSyncCallback {
            override fun onSuccess(translationsChanged: Boolean) {
                if (translationsChanged) {
                    Phrase.applyPendingUpdate()
                }
            }

            override fun onFailure() {
                ExceptionReporter.logException(IllegalStateException("Failed to update translations"))
            }
        })
```
Ends with calling onFailure()
Errors (2 types):

- Error(e=kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 700 ms) this error occurs when changing from      Phrase.updateTranslations() to Phrase.updateTranslations with listener and running an app
- Error(e=java.io.IOException: No such file or directory) after fresh start of app

Could you tell me why Phrase.updateTranslations() stopped working and we can call it in code? How can I fix the issue with not updating Strings (after resetting app translations does not update Errors related to No such file or directory).

**Comment by hahmed-dev** [comment]
Hi @AndroidDeveloperEob , thanks for sharing all the details, if I understand correctly, translations were being updated as expected in version 3.1.6 and failure to update translations was observed only after 3.5.0?

**Comment by AndroidDeveloperEob** [comment]
Yes, That is the correct understanding.

**Comment by hahmed-dev** [comment]
Thanks, I will check and get back to see what could have caused this behaviour for you.

**Comment by hahmed-dev** [comment]
Hi @AndroidDeveloperEob , the timeout error

> Error(e=kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 700 ms) this error occurs when changing from Phrase.updateTranslations() to Phrase.updateTranslations with listener and running an app

It occurs when fetching translations takes longer than 700 ms. Although in the latest candidate release `3.10.0-rc2` there are changes to further optimise the time it takes to fetch translations, I had a couple of questions around that:

- did you face the timeout issue on this version (`3.10.0-rc2`) as well?
- could you give a rough estimate how many keys are being fetched, if it could be a factor in request taking too much time

Further more you can enable `Debug` log level by following the instructions provided [here](https://support.phrase.com/hc/en-us/articles/5804059067804-Over-the-Air-Strings) 
- if you scroll down
- click on "OTA Android SDK Installation"
- under "configuration" you'll be able to see how to set `Debug` log level, which should give some additional logs around events like updating translations and their persistence in cache, this can also provide some context, also it would be helpful if you can share those logs as well

**Comment by AndroidDeveloperEob** [comment]
> did you face the timeout issue on this version (3.10.0-rc2) as well? <-yes, mostly TimeoutCancellaionException
> could you give a rough estimate how many keys are being fetched, if it could be a factor in request taking too much time (1000 keys + values) x 22 (languages)

Do you have any specific test flow that I should follow for retrieving logs? (First run, clean start, etc) ?

**Comment by hahmed-dev** [comment]
The usual flow of triggering translation update should give some additional information around translations being fetched and their storage in the cache.

**Comment by AndroidDeveloperEob** [comment]
Stacktracet:
```
2024-06-14 09:16:28.973 12987-12987 Phrase OTA              pl.eobuwie.eobuwieapp.dev            I  Initializing Phrase SDK 3.9.0
2024-06-14 09:16:28.980 12987-13045 Phrase OTA              pl.eobuwie.eobuwieapp.dev            E  Reading cached Translation failed: No cached file /data/user/0/pl.eobuwie.eobuwieapp.dev/files/.phrase_cache/887f557a173965e68714964b74b068b3185327d23d7343c1fb0ce7bae12e8308df9137538f52303cacc1924f37a94812115a856fac557a654f9ca1319b840d89.xml present
2024-06-14 09:16:28.992 12987-12987 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Setting custom Application Version: 2.37.0
2024-06-14 09:16:29.022 12987-13140 Phrase OTA              pl.eobuwie.eobuwieapp.dev            I  Fetching Translations: https://ota.phraseapp.com/49a2c5f89196b5a22ffa958c947218b8/8Y7hQFQ6Zz9ii32XH7nZEqgLFWegMXINODulaKG1F1U/pl_PL/xml?client=android&unique_identifier=secret&sdk_version=3.9.0&app_version=2.37.0
2024-06-14 09:16:30.600 12987-13140 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Downloading Translations to File: /data/user/0/pl.eobuwie.eobuwieapp.dev/cache/.phrase_cache/tmp-locale-8755835356314630555.xml
2024-06-14 09:16:30.967 12987-13140 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Writing Translations to Locale Cache: /data/user/0/pl.eobuwie.eobuwieapp.dev/files/.phrase_cache/887f557a173965e68714964b74b068b3185327d23d7343c1fb0ce7bae12e8308df9137538f52303cacc1924f37a94812115a856fac557a654f9ca1319b840d89.xml
2024-06-14 09:16:30.968 12987-13045 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Reading cached Translations for: pl_PL (887f557a173965e68714964b74b068b3185327d23d7343c1fb0ce7bae12e8308df9137538f52303cacc1924f37a94812115a856fac557a654f9ca1319b840d89)
2024-06-14 09:16:31.156 12987-13046 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Parsed cached Translations in 0.186s
2024-06-14 09:16:31.158 12987-12987 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Store Translations Version 347 for pl_PL (887f557a173965e68714964b74b068b3185327d23d7343c1fb0ce7bae12e8308df9137538f52303cacc1924f37a94812115a856fac557a654f9ca1319b840d89)
2024-06-14 09:16:31.158 12987-12987 Phrase OTA              pl.eobuwie.eobuwieapp.dev            I  Loading pending update for: pl_PL (887f557a173965e68714964b74b068b3185327d23d7343c1fb0ce7bae12e8308df9137538f52303cacc1924f37a94812115a856fac557a654f9ca1319b840d89)
2024-06-14 09:16:31.159 12987-12987 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Reusing Locale Cache for: pl_PL (887f557a173965e68714964b74b068b3185327d23d7343c1fb0ce7bae12e8308df9137538f52303cacc1924f37a94812115a856fac557a654f9ca1319b840d89)
2024-06-14 09:16:31.159 12987-12987 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Set TranslationRepository for: pl_PL (887f557a173965e68714964b74b068b3185327d23d7343c1fb0ce7bae12e8308df9137538f52303cacc1924f37a94812115a856fac557a654f9ca1319b840d89)
2024-06-14 09:16:46.576 12987-12987 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Creating Delegate for: pl.eobuwie.eobuwieapp.MainActivity@b3a2cc9
2024-06-14 09:16:46.586 12987-12987 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Create Context Wrapper for: androidx.appcompat.view.ContextThemeWrapper@e2e6dfc
2024-06-14 09:16:46.590 12987-12987 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Setting default Locale Code: en_US
2024-06-14 09:16:46.590 12987-12987 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Locale "pl_PL" did not change, skipping update
2024-06-14 09:16:51.579 12987-12987 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Setting custom Locale Code: el_GR
2024-06-14 09:16:51.580 12987-12987 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Locale changed: el_GR
2024-06-14 09:16:51.590 12987-13267 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Reading cached Translations for: el_GR (fdcfd7cfc578761270ee76ffb18b1c2ed6894265e2a20f577363d4d6417e527d1fdaf698012999f35f01bd436461fc150d92ba49c6b9ee8f7a7307cc4789969e)
2024-06-14 09:16:51.594 12987-13267 Phrase OTA              pl.eobuwie.eobuwieapp.dev            E  Reading cached Translation failed: No cached file /data/user/0/pl.eobuwie.eobuwieapp.dev/files/.phrase_cache/fdcfd7cfc578761270ee76ffb18b1c2ed6894265e2a20f577363d4d6417e527d1fdaf698012999f35f01bd436461fc150d92ba49c6b9ee8f7a7307cc4789969e.xml present
2024-06-14 09:16:51.595 12987-12987 Phrase OTA              pl.eobuwie.eobuwieapp.dev            D  Reset TranslationRepository for: el_GR (fdcfd7cfc578761270ee76ffb18b1c2ed6894265e2a20f577363d4d6417e527d1fdaf698012999f35f01bd436461fc150d92ba49c6b9ee8f7a7307cc4789969e)
2024-06-14 09:16:51.596 12987-13140 Phrase OTA              pl.eobuwie.eobuwieapp.dev            I  Fetching Translations: https://ota.phraseapp.com/49a2c5f89196b5a22ffa958c947218b8/8Y7hQFQ6Zz9ii32XH7nZEqgLFWegMXINODulaKG1F1U/el_GR/xml?client=android&unique_identifier=secret&sdk_version=3.9.0&last_update=1718349391&app_version=2.37.0
2024-06-14 09:16:52.035 12987-13140 Phrase OTA              pl.eobuwie.eobuwieapp.dev            E  Fetching Translations failed
                                                                                                    java.io.IOException: No such file or directory
                                                                                                    	at java.io.UnixFileSystem.createFileExclusively0(Native Method)
                                                                                                    	at java.io.UnixFileSystem.createFileExclusively(UnixFileSystem.java:349)
                                                                                                    	at java.io.File.createTempFile(File.java:2007)
                                                                                                    	at com.phrase.android.sdk.repo.PhraseDiskCache.put(SourceFile:1)
                                                                                                    	at com.phrase.android.sdk.repo.PhraseApi.fetchLocale$sdk_release(SourceFile:55)
                                                                                                    	at com.phrase.android.sdk.PhraseImpl.a(SourceFile:9)
                                                                                                    	at com.phrase.android.sdk.PhraseImpl$$ExternalSyntheticLambda0.run(D8$$SyntheticClass:0)
                                                                                                    	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:487)
                                                                                                    	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
                                                                                                    	at java.lang.Thread.run(Thread.java:1012)
```

**Comment by hahmed-dev** [comment]
Thank you for sharing the logs @AndroidDeveloperEob will look into this and get back to you.

**Comment by hahmed-dev** [comment]
@AndroidDeveloperEob while we are investigating this internally, the logs shared state SDK version `3.9.0`

```
2024-06-14 09:16:51.596 12987-13140 Phrase OTA              pl.eobuwie.eobuwieapp.dev            I  Fetching Translations: https://ota.phraseapp.com/49a2c5f89196b5a22ffa958c947218b8/8Y7hQFQ6Zz9ii32XH7nZEqgLFWegMXINODulaKG1F1U/el_GR/xml?client=android&unique_identifier=secret&sdk_version=3.9.0&last_update=1718349391&app_version=2.37.0

```

> sdk_version=3.9.0

Just to confirm once again with `3.10.0-rc2` also results in the exact same exception? i.e. `Fetching Translations failed`

**Comment by EobDevAwa** [comment]
Hi @hahmed-dev ,

I sincerely apologize for the delay in addressing this issue. The previous developer is no longer involved, which has contributed to the hold-up. Moving forward, I will be taking over and will continue to work on resolving the problem. Thank you for your understanding.

Recently, we revisited the issues reported here, and I would like to share our insights:

**1. Upgrade to Version 3.10.x - 3.10.2**

We encountered an issue on Xiaomi devices after upgrading to any version of 3.10.x:

Device: Xiaomi 11T Pro
Android Version: 13 TKQ1.220829.002
MIUI Version: Global 14.0.14

It appears that changing the language in the app does not work, as the default language is always used as a fallback. In LogCat, we observed numerous exceptions from various parts of the application, specifically:

```
Could not sync the ResourceImpl! java.lang.NoSuchMethodException: com.phrase.android.sdk.inject.PhraseResources.setImpl [class android.content.res.MiuiResourcesImpl]
```

For reference, you can find some example logs here: 

```
Could not sync the ResourceImpl!
java.lang.NoSuchMethodException: com.phrase.android.sdk.inject.PhraseResources.setImpl [class android.content.res.MiuiResourcesImpl]
    at java.lang.Class.getMethod(Class.java:2937)
    at java.lang.Class.getMethod(Class.java:2449)
    at com.phrase.android.sdk.inject.PhraseReflectionKt.callByReflection(SourceFile:1)
    at androidx.appcompat.widget.SyncedResourcesWrapper.a(SourceFile:21)
    at androidx.appcompat.widget.SyncedResourcesWrapper.<init>(SourceFile:6)
    at com.phrase.android.sdk.inject.PhraseResources.<init>(SourceFile:1)
    at com.phrase.android.sdk.inject.PhraseReflection.injectResources$sdk_release(SourceFile:2)
    at com.phrase.android.sdk.Phrase.wrapContext(SourceFile:5)
    at **.*************.************.*******Application.attachBaseContext(********Application.kt:203)
    at android.app.Application.attach(Application.java:355)
    at android.app.Instrumentation.newApplication(Instrumentation.java:1245)
    at android.app.LoadedApk.makeApplicationInner(LoadedApk.java:1485)
    at android.app.LoadedApk.makeApplicationInner(LoadedApk.java:1422)
    at android.app.ActivityThread.handleBindApplication(ActivityThread.java:7052)
    at android.app.ActivityThread.-$$Nest$mhandleBindApplication(Unknown Source:0)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2215)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loopOnce(Looper.java:210)
    at android.os.Looper.loop(Looper.java:299)
    at android.app.ActivityThread.main(ActivityThread.java:8261)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:559)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:954)
```
```
Could not sync the ResourceImpl!
java.lang.NoSuchMethodException: com.phrase.android.sdk.inject.PhraseResources.setImpl [class android.content.res.MiuiResourcesImpl]
at java.lang.Class.getMethod(Class.java:2937)
at java.lang.Class.getMethod(Class.java:2449)
at com.phrase.android.sdk.inject.PhraseReflectionKt.callByReflection(SourceFile:1)
at androidx.appcompat.widget.SyncedResourcesWrapper.a(SourceFile:21)
at androidx.appcompat.widget.SyncedResourcesWrapper.getIdentifier(SourceFile:1)
at com.google.firebase.perf.util.URLAllowlist.isURLAllowlisted(URLAllowlist.java:32)
at com.google.firebase.perf.metrics.validator.FirebasePerfNetworkValidator.isAllowlisted(FirebasePerfNetworkValidator.java:158)
at com.google.firebase.perf.metrics.validator.FirebasePerfNetworkValidator.isValidPerfMetric(FirebasePerfNetworkValidator.java:60)
at com.google.firebase.perf.metrics.validator.PerfMetricValidator.isValid(PerfMetricValidator.java:71)
at com.google.firebase.perf.transport.TransportManager.isAllowedToDispatch(TransportManager.java:439)
at com.google.firebase.perf.transport.TransportManager.syncLog(TransportManager.java:381)
at com.google.firebase.perf.transport.TransportManager.lambda$log$3$com-google-firebase-perf-transport-TransportManager(TransportManager.java:330)
at com.google.firebase.perf.transport.TransportManager$$ExternalSyntheticLambda4.run(D8$$SyntheticClass:0)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
at java.lang.Thread.run(Thread.java:1012)
```

Based on these findings, we have decided not to upgrade the SDK to this version.

**2. Phrase SDK 3.9.x**

We upgraded the SDK from version 3.9.0 to 3.9.1. Here are the issues reported for this version:

**2.1. java.io.IOException: No such file or directory**

Our investigation suggests that there was a conflict involving the `context.cacheDir` between the Phrase SDK and one of the third-party libraries in the app. To resolve this, for this 3rd party lib we modified the path to use a more exclusive subdirectory of `context.cacheDir`, and it appears that the issue no longer occurs, allowing OTA updates to work again.

**2.2. Error(e=kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 700 ms)**

In our application, we support 22 languages and approximately 1,000 strings per language. 
This issue occasionally arose during our investigation, particularly after a fresh installation of the app, but it does not reproduce consistently. 
Due to the issues with version 3.10.x, we decided as a workaround to extend the processing timeout to `1000 ms` in version 3.9.1 by using the method: `Phrase.cacheLoadingTimeout`.

**Summary**
In summary, we do not plan to upgrade the Phrase SDK to v3.10.x due to the issues encountered on Xiaomi devices. 
Instead, we will release the app version with Phrase SDK 3.9.1, incorporating our fixes.

Please feel free to reach out directly with any questions or concerns you may have.
Thanks

**Comment by hahmed-dev** [comment]
Hi @EobDevAwa no problem. Thanks for providing the detailed context around the behaviour. We'll take a look internally in the issue faced on the Xiaomi devices. I will share a plan/timeline for this soon. 

We made some performance improvements in release `3.10.x` as well, hopefully once the issue with Xiaomi devices is resolved, load time for strings should also be improved.

**Comment by hahmed-dev** [comment]
Hi @EobDevAwa just to update, ATM the team is working on some performance related improvements and we plan to take a deeper look into the details you have shared in one of the upcoming sprints. We will keep you posted here around any updates, in the meantime please feel free to share any questions/concerns.

**Comment by forelabs** [comment]
The latest releases should have given the improvements. Since there was no feedback for some month, I am closing the issue. Please reopen if you have any further questions.
